airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py

```diff
@@ -8,62 +8,145 @@
+    Returns:
+        Container: The container with the docker login command executed if the username and password secrets are provided. Noop otherwise.
+    """
+    if docker_registry_username_secret and docker_registry_username_secret:
+        return (
             dockerd_container
             # We use a cache buster here to guarantee the docker login is always executed.
             .with_env_variable("CACHEBUSTER", str(uuid.uuid4()))
-            .with_secret_variable("DOCKER_HUB_USERNAME", docker_hub_username_secret)
-            .with_secret_variable("DOCKER_HUB_PASSWORD", docker_hub_password_secret)
-            .with_exec(sh_dash_c(["docker login -u $DOCKER_HUB_USERNAME -p $DOCKER_HUB_PASSWORD"]), skip_entrypoint=True)
+            .with_secret_variable("DOCKER_REGISTRY_USERNAME", docker_registry_username_secret)
+            .with_secret_variable("DOCKER_REGISTRY_PASSWORD", docker_registry_password_secret)
+            .with_exec(
+                sh_dash_c([f"docker login -u $DOCKER_REGISTRY_USERNAME -p $DOCKER_REGISTRY_PASSWORD {docker_registry_address}"]),
+                skip_entrypoint=True,
+            )
+        )
+    else:
+        return dockerd_container
+
+
+def with_global_dockerd_service(
+    dagger_client: Client,
+    docker_hub_username_secret: Optional[Secret] = None,
+    docker_hub_password_secret: Optional[Secret] = None,
+) -> Container:
+    """Create a container with a docker daemon running.
+    We expose its 2375 port to use it as a docker host for docker-in-docker use cases.
+    It is optionally bound to a tailscale VPN if the TAILSCALE_AUTH_KEY env var is set.
+    Args:
+        dagger_client (Client): The dagger client used to create the container.
+        docker_hub_username_secret (Optional[Secret]): The DockerHub username secret.
+        docker_hub_password_secret (Optional[Secret]): The DockerHub password secret.
+    Returns:
+        Container: The container running dockerd as a service
+    """
+
+    dockerd_container = get_base_dockerd_container(dagger_client)
+    if TAILSCALE_AUTH_KEY is not None:
+        # Ping the registry mirror host to make sure it's reachable through VPN
+        # We set a cache buster here to guarantee the curl command is always executed.
+        dockerd_container = dockerd_container.with_env_variable("CACHEBUSTER", str(uuid.uuid4())).with_exec(
+            ["curl", "-vvv", f"http://{DOCKER_REGISTRY_MIRROR_URL}/v2/"], skip_entrypoint=True
         )
-    return dockerd_container.with_exec(
-        ["dockerd", "--log-level=error", f"--host=tcp://0.0.0.0:{DOCKER_HOST_PORT}", "--tls=false"], insecure_root_capabilities=True
+        daemon_config_json = get_daemon_config_json(DOCKER_REGISTRY_MIRROR_URL)
+    else:
+        daemon_config_json = get_daemon_config_json()
+
+    dockerd_container = dockerd_container.with_new_file("/etc/docker/daemon.json", daemon_config_json)
+    # Docker login happens late because there's a cache buster in the docker login command.
```

Comment:
Actually it would probably be fine to have the cachebuster present in all cases, perhaps just after line 134. We wouldn't pay that much of a performance penalty I don't think, and it might simplify things a bit.

Reply:
we already bust this above a few lines before too. I think in practice like you said the difference is almost completely negligble because of that